cloud/vsphere/machineactuator.go

```diff
@@ -506,69 +497,123 @@ func (vc *VsphereClient) Update(cluster *clusterv1.Cluster, goalMachine *cluster
 	}
 
 	if util.IsMaster(goalMachine) {
+		// Master upgrades
 		glog.Info("Upgrade for master machine.. Check if upgrade needed.")
-		// Get the versions for the new object.
-		goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
-		goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
 
 		// If the saved versions and new versions differ, do in-place upgrade.
-		if goalMachineControlPlaneVersion != goalMachine.Annotations[ControlPlaneVersionAnnotationKey] ||
-			goalMachineKubeletVersion != goalMachine.Annotations[KubeletVersionAnnotationKey] {
-			glog.Infof("Doing in-place upgrade for master from %s to %s", goalMachine.Annotations[ControlPlaneVersionAnnotationKey], goalMachineControlPlaneVersion)
+		if vc.needsMasterUpdate(goalMachine) {
+			glog.Infof("Doing in-place upgrade for master from v%s to v%s", goalMachine.Annotations[ControlPlaneVersionAnnotationKey], goalMachine.Spec.Versions.ControlPlane)
 			err := vc.updateMasterInPlace(goalMachine)
 			if err != nil {
-				glog.Errorf("Master inplace update failed: %+v", err)
+				glog.Errorf("Master in-place upgrade failed: %+v", err)
 			}
 		} else {
-			glog.Info("UPDATE TYPE NOT SUPPORTED")
+			glog.Info("UNSUPPORTED MASTER UPDATE.")
 			return nil
 		}
 	} else {
-		glog.Info("NODE UPDATES NOT SUPPORTED")
-		return nil
+		if vc.needsNodeUpdate(goalMachine) {
+			// Node upgrades
+			if err := vc.updateNode(cluster, goalMachine); err != nil {
+				return err
+			}
+		} else {
+			glog.Info("UNSUPPORTED NODE UPDATE.")
+			return nil
+		}
 	}
 
 	return nil
 }
 
-// Assumes that update is needed.
-// For now support only K8s control plane upgrades.
-func (vc *VsphereClient) updateMasterInPlace(goalMachine *clusterv1.Machine) error {
-	goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
-	goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
+func (vc *VsphereClient) needsControlPlaneUpdate(machine *clusterv1.Machine) bool {
+	return machine.Spec.Versions.ControlPlane != machine.Annotations[ControlPlaneVersionAnnotationKey]
+}
+
+func (vc *VsphereClient) needsKubeletUpdate(machine *clusterv1.Machine) bool {
+	return machine.Spec.Versions.Kubelet != machine.Annotations[KubeletVersionAnnotationKey]
+}
 
-	currentControlPlaneVersion := goalMachine.Annotations[ControlPlaneVersionAnnotationKey]
-	currentKubeletVersion := goalMachine.Annotations[KubeletVersionAnnotationKey]
+// Returns true if the node is needed to be upgraded.
+func (vc *VsphereClient) needsNodeUpdate(machine *clusterv1.Machine) bool {
+	return !util.IsMaster(machine) &&
+		vc.needsKubeletUpdate(machine)
+}
 
-	// Control plane upgrade
-	if goalMachineControlPlaneVersion != currentControlPlaneVersion {
+// Returns true if the master is needed to be upgraded.
+func (vc *VsphereClient) needsMasterUpdate(machine *clusterv1.Machine) bool {
+	return util.IsMaster(machine) &&
```

Comment:
This is only called from within a block that already checks `util.IsMaster`. Is this just here for extra safety or is it from before the call was moved inside that block? The function is private and the name implies that it should only be used for masters.